File: rd/Client/cms/Assets/StreamingAssets/lua/logic/SpellService/EffectApplyBuff.lua
local string = string
local math = math
local pairs = pairs
local setmetatable = setmetatable
local table = table
local Effect = Effect
local _G = _G
local Log = DebugLog

module "EffectApplyBuff"
---------------------------------------------------------------------------------------------------
EffectApplyBuff = {
	buffID,
	--buffCount = 1,
}
setmetatable(EffectApplyBuff, {__index = Effect.Effect})

---------------------------------------------------------------------------------------------------
function EffectApplyBuff:Apply(curTime)
	if self:ApplyBasic(curTime) == false then
		do return end
	end
	
	local info = string.format("apply buff effect Apply id%s \n", self.id)
	Log:Log(info)
	
	self:CalculateHit()
	
	local buff = self.owner:GetBuff(self.buffID)
	if self.hit == _G.HIT_SUCCESS then
		if buff ~= nil then
			buff.caster = self.caster
			buff.target = self.target
			buff:SetOwnedSpell(self.ownedSpell)
			buff:Apply(self.applyTime)
			--self:CalculateEnergy()
		end
	end
end
---------------------------------------------------------------------------------------------------
function EffectApplyBuff:CalculateHit()
	--check exclusion
	local activeBuffList = self.target.buffList
	if activeBuffList then
		for _, buff in pairs(activeBuffList) do
			if not(buff.isFinish) then
				if buff.buffExclusionList then
					for _, excludeID in pairs( buff.buffExclusionList) do
						if excludeID == self.buffID then
							self.hit = _G.HIT_IMMUNE
							return
						end
					end
				end
				if buff.buffCategoryExclusionList then
					local targetBuff = self.owner:GetBuff(self.buffID)
					if buff then
						for _, excludeCategory in pairs( buff.buffCategoryExclusionList) do
							if targetBuff.category[excludeCategory] then
								self.hit = _G.HIT_IMMUNE
								return
							end
						end
					end
				end
			end
		end
	end
	
	--not check ally team
	if _G.IsAlly(self) then
		self.hit = _G.HIT_SUCCESS
		return
	end
	
	self.hit = self:CalculateBasicHit(self.caster, self.target)
end
---------------------------------------------------------------------------------------------------

File: rd/trunk/Client/cms/Assets/script/SpellService/EffectApplyBuff.cs
using UnityEngine;
using System;
using System.Collections;


[Serializable]
public class EffectApplyBuffPrototype : EffectPrototype
{
    public string buffID;
    public int validatorNum;
}

public class EffectApplyBuff : Effect
{
    //---------------------------------------------------------------------------------------------
    public override void Init(EffectPrototype pt, SpellService owner)
    {
        EffectApplyBuffPrototype buffPtOut = pt as EffectApplyBuffPrototype;
        protoEffect = new EffectApplyBuffPrototype();
        EffectApplyBuffPrototype  buffPt = protoEffect as EffectApplyBuffPrototype;
        buffPt.buffID = buffPtOut.buffID;
        buffPt.validatorNum = buffPtOut.validatorNum;
        base.Init(pt, owner);
    }
    //---------------------------------------------------------------------------------------------
    public override bool Apply(float applyTime, string wpName)
    {
        if (base.Apply(applyTime, wpName) == false)
            return false;

        int hitResult = CalculateHit(wpName);

        Logger.Log("[SpellService]trigger apply buff effect");
        if (hitResult == SpellConst.hitSuccess)
        {
            EffectApplyBuffPrototype applyBuffProt = protoEffect as EffectApplyBuffPrototype;
            if (applyBuffProt == null)
            {
                return false;
            }
            Buff curBuff = spellService.GetBuff(applyBuffProt.buffID);
            if (curBuff != null)
            {
                GameUnit caster = spellService.GetUnit(casterID);
                int buffCount = caster.buffList.Count;
                for (int i = 0; i < buffCount; ++i)
                {
                    caster.buffList[i].DamageResponse(applyTime, this, wpName);
                }
                curBuff.casterID = casterID;
                curBuff.targetID = targetID;
                curBuff.SetOwnedSpell(ownedSpell);
                curBuff.Apply(applyTime);
            }

            //link effect
            Effect curEffect = spellService.GetEffect(protoEffect.linkEffect);
            if (curEffect != null)
            {
                curEffect.SetOwnedBuff(ownedBuff);
                curEffect.SetOwnedSpell(ownedSpell);
                curEffect.targetID = targetID;
                curEffect.absoluteHit = true;
                curEffect.Apply(applyTime, wpName);
            }
        }

        return true;
    }
    //---------------------------------------------------------------------------------------------
    public override int CalculateHit(string wpName)
    {
        GameUnit target = spellService.GetUnit(targetID);
        //check prop first
        EffectApplyBuffPrototype buffPt = protoEffect as EffectApplyBuffPrototype;
        if (buffPt != null && buffPt.validatorNum > 0)
        {
            int targetProp = 1 << (target.property - 1);
            if ((targetProp & buffPt.validatorNum) == 0)
            {
                return SpellConst.hitIgnore;
            }
        }

        if (absoluteHit == true || casterID == BattleConst.battleSceneGuid)
            return SpellConst.hitSuccess;
        //TODO:
        //1 check buff replace

        //2 check ally team
        GameUnit caster = spellService.GetUnit(casterID);
        if (caster.pbUnit.camp == target.pbUnit.camp)
        {
            return SpellConst.hitSuccess;
        }

        //check is weak point damagepoint
        WeakPointRuntimeData wpRuntime = null;
        if (string.IsNullOrEmpty(wpName) == false)
        {
            target.battleUnit.wpGroup.allWpDic.TryGetValue(wpName, out wpRuntime);
            if (wpRuntime != null && wpRuntime.staticData.isDamagePoint != 1)
            {
                EffectApplyBuffPrototype applyBuffProt = protoEffect as EffectApplyBuffPrototype;
                if (applyBuffProt != null)
                {
                    Buff curBuff = spellService.GetBuff(applyBuffProt.buffID);
                    if (curBuff != null && curBuff.buffProto.category == (int)BuffType.Buff_Type_Dot)
                    {
                        //miss event
                        SpellVitalChangeArgs args = new SpellVitalChangeArgs();
                        args.vitalType = (int)VitalType.Vital_Type_Immune;
                        args.triggerTime = applyTime;
                        args.casterID = casterID;
                        args.targetID = targetID;
                        spellService.TriggerEvent(GameEventList.SpellLifeChange, args);
                        return SpellConst.hitImmune;
                    }
                }
            }
        }

       return base.CalculateHit(wpName);
    }
    //--------------------------------------------------------------------------------------------- 
}
